Add a Spark Kryo serializer for SAMRecords.

tomwhite opened the pull request:
This improves the serialized size considerably, which has a very big effect on the amount of data going through the shuffle.

* SAMRecordToGATKReadAdapterSerializer uses BAMRecordCodec for a compact serialization format.
* Record headers are not needed since GATKRead is intended to abstract them away. Note that the reference name is serialized separately so it can be preserved.
* ReadsPreprocessingPipelineSparkTestData creates reads with zero UUIDs since they are not used in Spark.

tomwhite commented:
@davidadamsphd could review this when you have time please?

davidadamsphd commented on src/test/java/org/broadinstitute/hellbender/engine/spark/ReadsPreprocessingPipelineSparkTestData.java:
```diff
@@ -0,0 +1,205 @@
+ * ReadsPreprocessingPipelineTestData contains coordinated test data that can be used in the many transforms that
+ * are a part of the ReadsPreprocessingPipeline.
+ */
+public class ReadsPreprocessingPipelineSparkTestData {
```
What's changed here relative to ReadsPreprocessingPipelineTestData?

tomwhite replied:
The UUIDs are alway 0 for Spark, otherwise comparison of reads always fails.

davidadamsphd commented on src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java:
```diff
@@ -103,7 +98,12 @@ private static boolean samRecordOverlaps(final SAMRecord record, final List<Simp
         }
         for (SimpleInterval interval : intervals) {
-            if (interval.overlaps(record)) {
+            if (record.getReadUnmappedFlag() && record.getAlignmentStart() != SAMRecord.NO_ALIGNMENT_START) {
```
I want to also run this by @droazen , to make sure it looks right to him (I don't know the subtitles of unmapped reads well).

tomwhite replied:
I think it's OK as he merged the equivalent code into dataflow a couple of days ago: https://github.com/broadinstitute/hellbender/pull/880.

davidadamsphd replied:
Agreed.

davidadamsphd commented:
This is great!
However, I have a two concerns:
(1) There are some functions that can be called on `SAMRecordToGATKReadAdapter` that cause the header to be queried (or produce different results if the header is missing). This is something that @jean-philippe-martin  and @droazen  noticed.

We definitely want to avoid shipping around the header if at all possible (which I think we can do), but we need to solve _that_ problem before we turn this on.

(2) We don't have sufficient testing in place to catch the issue from (1).
I propose that I create a ticket on getting rid of the header and we discuss the possible solutions for that problem over there. For this PR, I think it should go forward but without hooking it up to `GATKRegistrator `.  You can write (copy) a unit test that hooks it up and shows that it works. In the mean time, we're collectively porting mark dupes and BQSR/ApplyBQSR. Once we have enough in place to have some confidence that we'd catch subtle serialization bugs (like missing headers), we should hook it up.

davidadamsphd commented:
I created #900 to discuss the header issue.

jean-philippe-martin commented:
We should band together. The Dataflow side would also benefit from a better serializer.

lbergelson commented:
I tested this against the spark BQSR and everything seems to be working so
long as we add an unmapped read filter to the pipeline.

It looks really good. Seems to be much faster. The variants are killing us
at the moment. I added a stripped down variant context and runtime on my
test input goes from 40 minutes to 2.
On Sep 16, 2015 8:05 PM, "JP Martin" <notifications@github.com> wrote:

> We should band together. The Dataflow side would also benefit from a
> better serializer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/hellbender/pull/899#issuecomment-140929673>
> .
>

laserson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java:
```diff
@@ -0,0 +1,97 @@
+
+        // set reference indexes again since setting reference names without a header will unset indexes
+        record.setReferenceIndex(referenceIndex);
+        record.setMateReferenceIndex(mateReferenceIndex);
```
Not sure these are necessary as setting the reference name should also set the reference index.  And retrieving the reference index will certainly evaluate it if it's unset.

tomwhite replied:
This dance is needed due to the fact that we don't have a header so we have to set the reference names and indexes explicitly. This ordering ensures that they all get set safely.

laserson replied:
Ah, I see.

laserson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java:
```diff
@@ -0,0 +1,97 @@
+        lazyCodec.encode(record);
+
+        // clear indexing bin after encoding to ensure all SAMRecords compare properly
+        record.setFlags(record.getFlags());
```
Why do you have to mess with the `SAMRecord` object here after it's already been encoded into the byte stream?

tomwhite replied:
Because calling encode will mutate the record (set the indexing bin field) which has implications for equality.

laserson commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SAMRecordToGATKReadAdapterSerializer.java:
```diff
@@ -0,0 +1,97 @@
+        record.setFlags(record.getFlags());
+
+        final int referenceIndex = record.getReferenceIndex();
+        final int mateReferenceIndex = record.getMateReferenceIndex();
```
These two also not needed I think.

tomwhite commented:
@davidadamsphd, I've followed your suggestion and removed the serializer from the registrator until we've properly addressed #900. I've added a unit test for the serializer (just for the serializer itself, since having a different registrator is problematic given we have a shared spark context for all tests).

As a check I ran the Spark mark duplicates code with this change and the tests passed. I've also sorted a large BAM with #901 and this change and it worked well.

@jean-philippe-martin, definitely! The serialization code should carry over to dataflow coders.

@lbergelson, yes, we absolutely want to scrutinize all the types that go through the shuffle to make sure they are reasonably compact.

davidadamsphd commented on src/test/java/org/broadinstitute/hellbender/engine/spark/ReadsPreprocessingPipelineSparkTestData.java:
```diff
@@ -0,0 +1,205 @@
+    private final List<Variant> variants;
+    private final List<KV<GATKRead, ReferenceBases>> kvReadsRefBases;
+    private final List<KV<GATKRead, Variant>> kvReadVariant;
+    private final List<KV<GATKRead, Iterable<Variant>>> kvReadiVariantBroken; // The dataflow version is currently broken (Issue #795).
```
You should be able to remove this (it's just for Dataflow).

davidadamsphd commented:
LGTM, modulo the nit.

laserson commented:
+1
